File: tests/worker/request-history.mjs
import assert from 'node:assert/strict';
import { readFileSync, readdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { Miniflare, Response, Log, LogLevel } from 'miniflare';

const mf = new Miniflare({
  log: new Log(LogLevel.ERROR), modules: true,
  scriptPath: fileURLToPath(new URL('./worker.mjs', import.meta.url)),
  modulesRules: [{ type: 'CompiledWasm', include: ['**/*.wasm'] }],
  compatibilityDate: '2024-09-23', compatibilityFlags: ['nodejs_compat'],
  kvNamespaces: ['SESSIONS'], d1Databases: ['DB'],
  bindings: {
    GHINVITE_BASE_URL: 'https://history.test', GHINVITE_SESSION_SECRET: '07'.repeat(32),
    GHINVITE_RESTATE_INGRESS: 'https://restate.test', GHINVITE_RESTATE_AUTH: 'local-unauthenticated',
    GHINVITE_GITHUB_INSTALL_URL: 'https://github.com/apps/dummy/installations/new',
    GHINVITE_GITHUB_CLIENT_ID: 'dummy', GHINVITE_GITHUB_CLIENT_SECRET: 'dummy', GHINVITE_WEBHOOK_SECRET: 'dummy',
  },
  async outboundService(request) {
    if (request.url === 'https://github.com/login/oauth/access_token') return Response.json({ access_token: 'fixture', token_type: 'bearer', scope: '' });
    if (request.url === 'https://api.github.com/user') return Response.json({ id: 42, login: 'octocat', avatar_url: null });
    throw new Error(`unexpected outbound ${request.url}`);
  },
});

try {
  const db = await mf.getD1Database('DB');
  for (const file of readdirSync(new URL('../../migrations/', import.meta.url)).filter(f => f.endsWith('.sql')).sort()) {
    await db.exec(readFileSync(new URL(`../../migrations/${file}`, import.meta.url), 'utf8').replace(/^--.*$/gm, '').replaceAll('\n', ' '));
  }
  await db.prepare("INSERT INTO installations VALUES (1,42,'octocat','User','2026-01-01T00:00:00Z',NULL,'[]')").run();
  await db.prepare("INSERT INTO users VALUES (42,'octocat',NULL,'2026-01-01T00:00:00Z')").run();
  const link = '01ARZ3NDEKTSV4RRFFQ69G5FAV';
  const foreign = '01ARZ3NDEKTSV4RRFFQ69G5FAW';
  for (const [id, account, code] of [[link, 42, 'HistoryCode00001'], [foreign, 999, 'ForeignCode00001']]) {
    await db.prepare(`INSERT INTO invitation_links (id,slug,installation_id,account_id,created_by,created_at,permission,approval_required,description)
      VALUES (?,?,1,?,42,'2026-01-01T00:00:00Z','pull',1,'History fixture')`).bind(id, code, account).run();
    await db.prepare("INSERT INTO invitation_link_repos VALUES (?,10,'octocat/api')").bind(id).run();
  }
  const login = await mf.dispatchFetch('https://history.test/login', { redirect: 'manual' });
  const state = new URL(login.headers.get('location')).searchParams.get('state');
  let cookie = login.headers.get('set-cookie').split(';')[0];
  const callback = await mf.dispatchFetch(`https://history.test/oauth/callback?code=dummy&state=${state}`, { headers: { cookie }, redirect: 'manual' });
  assert.equal(callback.status, 303);
  cookie = callback.headers.get('set-cookie').split(';')[0];
  const get = async (path, status = 200) => {
    const response = await mf.dispatchFetch(`https://history.test${path}`, { headers: { cookie } });
    assert.equal(response.status, status, await response.clone().text());
    assert.equal(response.headers.get('cache-control'), 'private, no-store');
    return response.text();
  };
  const base = `/console/accounts/octocat/links/${link}/requests`;
  assert.match(await get(base), /No projected requests in this range/);
  const ids = [];
  const alphabet = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
  for (let n = 0; n < 53; n++) {
    ids[n] = '0'.repeat(24) + alphabet[Math.floor(n / 32)] + alphabet[n % 32];
    await db.prepare(`INSERT INTO invitation_requests (id,invitation_link_id,requester_id,state,created_at,decided_at,decline_reason)
      VALUES (?,?,42,'declined',?,'2026-01-02T00:00:00Z','private decision')`)
      .bind(ids[n], n === 52 ? foreign : link, n === 51 ? '2026-01-01T00:00:00.000000001Z' : n % 2 ? '2026-01-01T00:00:00+00:00' : '2026-01-01T00:00:00Z').run();
  }
  const found = [];
  let path = base;
  for (const size of [25, 25, 2]) {
    const html = await get(path);
    const page = [...html.matchAll(/href="\/console\/accounts\/octocat\/requests\/([A-Z0-9]+)"/g)].map(m => m[1]);
    assert.equal(page.length, size);
    found.push(...page);
    path = html.match(/href="([^"]+\?before=[^"]+)"/)?.[1];
  }
  assert.equal(path, undefined);
  assert.deepEqual(found, ids.slice(0, 52).reverse());
  const detail = `/console/accounts/octocat/requests/${ids[51]}`;
  let html = await get(detail);
  assert.match(html, /private decision/);
  assert.match(html, /Immutable requested permission: pull/);
  assert.match(html, /Projected history/);
  assert.doesNotMatch(await get(`/console/accounts/octocat/requests/${ids[52]}`, 404), /private decision/);
  await get(`/console/accounts/octocat/links/${foreign}/requests`, 404);
  await db.prepare("UPDATE users SET last_seen_at='broken'").run();
  assert.match(await get(detail), /GitHub user ID 42 · Profile unavailable/);
  await db.exec('DROP TABLE delivery_outcomes');
  assert.match(await get(detail), /Delivery information unavailable/);
  await db.prepare("UPDATE invitation_requests SET state='bad' WHERE id=?").bind(ids[51]).run();
  assert.match(await get(base, 503), /Request history unavailable/);
  console.log('PASS D1 request history: bounded pages, timestamp precision, terminal details, foreign concealment, missing profile and unavailable reads');
} finally {
  await mf.dispose();
}

File: crates/ghinvite-web/tests/console_flow/request_history.rs
use super::*;
use ghinvite_core::storage::projection::fixture::Seed;
use ghinvite_core::storage::{DeliveryStorage, InstallationStorage, RecordStorage};

#[tokio::test]
async fn history_and_detail_require_current_account_admin_authority() {
    let storage = Arc::new(
        ghinvite_storage_sqlx::SqlxStorage::in_memory()
            .await
            .unwrap(),
    );
    let (app, cookie) = links_app(storage.clone(), "member").await;
    let link = list_link(1);
    storage.seed_link(&link).await.unwrap();
    for path in [
        format!("/console/accounts/acme/links/{}/requests", link.id),
        format!(
            "/console/accounts/acme/requests/{}",
            ghinvite_core::RequestId::new()
        ),
    ] {
        let response = identity_request(&app, &cookie, "GET", &path).await;
        assert_eq!(response.status(), StatusCode::NOT_FOUND);
        assert!(!response_html(response).await.contains("Workshop"));
        let response = identity_request(&app, "", "GET", &path).await;
        assert_eq!(response.status(), StatusCode::SEE_OTHER);
        assert!(
            response.headers()["location"]
                .to_str()
                .unwrap()
                .contains("/login")
        );
    }
}

#[tokio::test]
#[ignore = "long-running Playwright fixture"]
async fn request_history_browser_server() {
    use axum::response::IntoResponse;
    let (app, cookie, storage, envelope) = deadline_queue_app(None).await;
    let mut request = storage
        .get_invitation_request(envelope.requests[0].request_id)
        .await
        .unwrap()
        .unwrap();
    request.state = ghinvite_core::RequestState::Declined;
    request.decided_by = Some(42);
    request.decided_at = Some("2026-01-02T12:00:00Z".parse().unwrap());
    request.decline_reason = Some("Browser decision".into());
    for _ in 0..27 {
        request.id = ghinvite_core::RequestId::new();
        storage.seed_request(&request).await.unwrap();
    }
    let destination = format!("/console/accounts/octocat/links/{}", envelope.link.link_id);
    let fixture = axum::Router::new()
        .route(
            "/fixture-login",
            axum::routing::get(move || {
                let cookie = cookie.clone();
                let destination = destination.clone();
                async move {
                    (
                        [(
                            axum::http::header::SET_COOKIE,
                            format!("{cookie}; Path=/; HttpOnly; SameSite=Lax"),
                        )],
                        axum::response::Redirect::to(&destination),
                    )
                        .into_response()
                }
            }),
        )
        .route("/health", axum::routing::get(|| async { StatusCode::OK }))
        .fallback(move |request: Request<Body>| {
            let app = app.clone();
            async move { app.oneshot(request).await.unwrap() }
        });
    let listener = tokio::net::TcpListener::bind("127.0.0.1:4178")
        .await
        .unwrap();
    axum::serve(listener, fixture).await.unwrap();
}

#[tokio::test]
async fn request_history_handles_empty_missing_profiles_and_unavailable_projections() {
    let path = std::env::temp_dir().join(format!(
        "ghinvite-history-{}.sqlite",
        ghinvite_core::RequestId::new()
    ));
    let storage = Arc::new(
        ghinvite_storage_sqlx::SqlxStorage::at_path(&path)
            .await
            .unwrap(),
    );
    let (app, cookie) = links_app(storage.clone(), "admin").await;
    let mut link = list_link(1);
    link.repos = vec![ghinvite_core::InvitationLinkRepo {
        repo_id: 10,
        repo_full_name: "acme/api".into(),
    }];
    storage.seed_link(&link).await.unwrap();
    let history = format!("/console/accounts/acme/links/{}/requests", link.id);
    let html = response_html(identity_request(&app, &cookie, "GET", &history).await).await;
    assert!(html.contains("No projected requests in this range"));
    let request = ghinvite_core::InvitationRequest {
        id: ghinvite_core::RequestId::new(),
        invitation_link_id: link.id,
        requester_id: 42,
        justification: Some("private justification".into()),
        state: ghinvite_core::RequestState::Pending,
        decided_by: None,
        decided_at: None,
        decline_reason: None,
        decision_deadline: Some("2026-01-03T12:34:56Z".parse().unwrap()),
        created_at: "2026-01-01T00:00:00Z".parse().unwrap(),
    };
    storage.seed_request(&request).await.unwrap();
    let detail = format!("/console/accounts/acme/requests/{}", request.id);
    let pool =
        sqlx::SqlitePool::connect_with(sqlx::sqlite::SqliteConnectOptions::new().filename(&path))
            .await
            .unwrap();
    // A missing optional profile must preserve the immutable identity.
    sqlx::query("UPDATE users SET last_seen_at = 'broken'")
        .execute(&pool)
        .await
        .unwrap();
    let html = response_html(identity_request(&app, &cookie, "GET", &detail).await).await;
    assert!(html.contains("GitHub user ID 42 · Profile unavailable"));
    assert!(html.contains("Decision deadline passed. The projected state may be delayed"));
    sqlx::query("DROP TABLE delivery_outcomes")
        .execute(&pool)
        .await
        .unwrap();
    let html = response_html(identity_request(&app, &cookie, "GET", &detail).await).await;
    assert!(html.contains("Delivery information unavailable"));
    assert!(html.contains("Current delivery status unavailable"));
    assert!(!html.contains("No projected delivery outcome."));
    assert!(html.contains("private justification"));
    sqlx::query("UPDATE invitation_requests SET state = 'corrupt'")
        .execute(&pool)
        .await
        .unwrap();
    let response = identity_request(&app, &cookie, "GET", &history).await;
    assert_eq!(response.status(), StatusCode::SERVICE_UNAVAILABLE);
    let html = response_html(response).await;
    assert!(html.contains("Request history unavailable"));
    assert!(!html.contains("No projected requests"));
    sqlx::query("UPDATE invitation_requests SET state = 'pending'")
        .execute(&pool)
        .await
        .unwrap();
    let mut connection = pool.acquire().await.unwrap();
    sqlx::query("PRAGMA foreign_keys = OFF")
        .execute(&mut *connection)
        .await
        .unwrap();
    sqlx::query("DELETE FROM invitation_links")
        .execute(&mut *connection)
        .await
        .unwrap();
    drop(connection);
    let response = identity_request(&app, &cookie, "GET", &detail).await;
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
    assert!(
        !response_html(response)
            .await
            .contains("private justification")
    );
    pool.close().await;
    drop(app);
    drop(storage);
    std::fs::remove_file(path).unwrap();
}

#[tokio::test]
async fn request_history_pages_and_terminal_details_conceal_foreign_resources() {
    let (app, cookie, storage, envelope) = deadline_queue_app(None).await;
    let link = envelope.link.link_id;
    let mut template = storage
        .get_invitation_request(envelope.requests[0].request_id)
        .await
        .unwrap()
        .unwrap();
    template.state = ghinvite_core::RequestState::Cancelled;
    for _ in 0..27 {
        template.id = ghinvite_core::RequestId::new();
        storage.seed_request(&template).await.unwrap();
    }
    let base = format!("/console/accounts/octocat/links/{link}/requests");
    let html = response_html(identity_request(&app, &cookie, "GET", &base).await).await;
    assert_eq!(html.matches("State:").count(), 25);
    let older = html
        .split("href=\"")
        .filter_map(|s| s.split('"').next())
        .find(|s| s.contains("?before="))
        .unwrap();
    let page = response_html(identity_request(&app, &cookie, "GET", older).await).await;
    assert_eq!(page.matches("State:").count(), 3);
    assert!(page.contains("Latest requests"));
    assert!(!page.contains("Older requests"));
    let malformed =
        response_html(identity_request(&app, &cookie, "GET", &format!("{base}?before=bad")).await)
            .await;
    assert_eq!(malformed.matches("State:").count(), 25);
    let mut foreign = storage
        .get_invitation_link_by_id(link)
        .await
        .unwrap()
        .unwrap();
    foreign.id = ghinvite_core::InvitationLinkId::new();
    foreign.account_id = 999;
    foreign.installation_id = 999;
    storage
        .insert_installation(&ghinvite_core::Account {
            installation_id: 999,
            account_id: 999,
            account_login: "foreign".into(),
            account_type: ghinvite_core::AccountType::Organization,
            installed_at: foreign.created_at,
            uninstalled_at: None,
            selected_repos: ghinvite_core::SelectedRepos::All,
        })
        .await
        .unwrap();
    foreign.slug = ghinvite_core::Slug::from_string("ForeignHistory01".into()).unwrap();
    foreign.description = "Foreign private link".into();
    storage.seed_link(&foreign).await.unwrap();
    template.id = ghinvite_core::RequestId::new();
    template.invitation_link_id = foreign.id;
    template.justification = Some("Foreign private justification".into());
    storage.seed_request(&template).await.unwrap();
    for path in [
        format!("/console/accounts/octocat/links/{}/requests", foreign.id),
        format!("/console/accounts/octocat/requests/{}", template.id),
    ] {
        let response = identity_request(&app, &cookie, "GET", &path).await;
        assert_eq!(response.status(), StatusCode::NOT_FOUND);
        let html = response_html(response).await;
        assert!(!html.contains("Foreign private"));
        assert!(!html.contains("octocat/api"));
    }
}

#[tokio::test]
async fn approved_request_distinguishes_projected_outcomes_from_missing_delivery() {
    use ghinvite_core::delivery::{CreateCommand, CreateOutcome, CreateReceipt};
    let (app, cookie, storage, envelope) = deadline_queue_app(None).await;
    let request = &envelope.requests[0];
    {
        let mut decided = storage
            .get_invitation_request(request.request_id)
            .await
            .unwrap()
            .unwrap();
        decided.state = ghinvite_core::RequestState::Approved;
        decided.decided_by = Some(42);
        decided.decided_at = Some("2026-01-02T12:00:00Z".parse().unwrap());
        decided.decline_reason = None;
        storage.seed_decision(&decided).await.unwrap();
    }
    let path = format!("/console/accounts/octocat/requests/{}", request.request_id);
    let html = response_html(identity_request(&app, &cookie, "GET", &path).await).await;
    assert!(html.contains("No projected delivery outcome yet"));
    let mut receipt = CreateReceipt {
        command: CreateCommand {
            invitation_id: ghinvite_core::GithubInvitationId::new(),
            link_id: request.link_id,
            request_id: request.request_id,
            approval_id: "approval".into(),
            account_id: 42,
            installation_id: 77,
            requester_id: 99,
            repo_id: 10,
            repo_full_name: "octocat/api".into(),
            permission: ghinvite_core::Permission::Pull,
            approved_at: "2026-01-02T12:00:00Z".parse().unwrap(),
        },
        outcome: CreateOutcome::Blocked {
            reason: "private infrastructure diagnostic".into(),
        },
        revision: 1,
        confirmed_at: None,
    };
    for (outcome, label) in [
        (
            receipt.outcome.clone(),
            "Blocked — waiting for availability or identity verification",
        ),
        (
            CreateOutcome::OutcomeUnknown,
            "GitHub outcome unknown — awaiting reconciliation",
        ),
        (
            CreateOutcome::Created { upstream_id: 12345 },
            "GitHub invitation created — awaiting acceptance",
        ),
        (CreateOutcome::AlreadyCollaborator, "Already a collaborator"),
        (
            CreateOutcome::Failed { status: 422 },
            "GitHub rejected delivery",
        ),
    ] {
        receipt.outcome = outcome;
        receipt.revision += 1;
        storage.project_delivery(&receipt).await.unwrap();
        let html = response_html(identity_request(&app, &cookie, "GET", &path).await).await;
        assert!(html.contains(label), "missing {label}");
        if matches!(receipt.outcome, CreateOutcome::Created { .. }) {
            assert!(html.contains("Current delivery status unavailable"));
            assert!(html.contains("retained history"));
        }
        assert!(!html.contains("private infrastructure diagnostic"));
    }
}

#[tokio::test]
async fn terminal_request_is_reachable_from_link_history_with_immutable_scope_and_decision() {
    let (app, cookie, storage, envelope) = deadline_queue_app(None).await;
    let id = envelope.requests[0].request_id;
    let link = envelope.link.link_id;
    {
        let mut decided = storage.get_invitation_request(id).await.unwrap().unwrap();
        decided.state = ghinvite_core::RequestState::Declined;
        decided.decided_by = Some(42);
        decided.decided_at = Some("2026-01-02T12:00:00Z".parse().unwrap());
        decided.decline_reason = Some("Admin-only decision context".into());
        storage.seed_decision(&decided).await.unwrap();
    }
    let base = "/console/accounts/octocat";
    let html = response_html(
        identity_request(&app, &cookie, "GET", &format!("{base}/links/{link}")).await,
    )
    .await;
    assert!(html.contains(&format!("{base}/links/{link}/requests")));
    let response = identity_request(
        &app,
        &cookie,
        "GET",
        &format!("{base}/links/{link}/requests"),
    )
    .await;
    assert_eq!(response.status(), StatusCode::OK);
    let html = response_html(response).await;
    assert!(html.contains("Request history"));
    assert!(html.contains("@requester"));
    assert!(html.contains("declined"));
    assert!(html.contains(&format!("{base}/requests/{id}")));
    let response = identity_request(&app, &cookie, "GET", &format!("{base}/requests/{id}")).await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_eq!(response.headers()["cache-control"], "private, no-store");
    let html = response_html(response).await;
    for expected in [
        "Invitation request",
        "declined",
        "octocat/api",
        "pull",
        "Admin-only decision context",
        "2026-01-02 12:00:00 UTC",
        "GitHub user ID 42",
        "Projected history",
        "Decision deadline",
    ] {
        assert!(html.contains(expected), "missing {expected}: {html}");
    }
}
